sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala

```diff
@@ -57,26 +90,135 @@ case class GpuNot(child: Expression) extends CudfUnaryExpression
+          val lhsNoNulls = withResource(Scalar.fromBool(true)) { trueScalar =>
+            lhsBool.getBase.replaceNulls(trueScalar)
+          }
+          withResource(lhsNoNulls) { _ =>
+            val rEval = withResource(filterBatch(tbl, lhsNoNulls, colTypes)) { leftTrueBatch =>
+              rightExpr.columnarEval(leftTrueBatch)
+            }
+            val finalRet = withResourceIfAllowed(rEval) { _ =>
+              (rEval) match {
+                case (f: GpuColumnVector) =>
+                  withResource(gather(lhsNoNulls, f)) { combinedVector =>
+                    doColumnar(lhsBool, GpuColumnVector.from(combinedVector, dataType))
+                  }
+                case (f: GpuScalar) =>
+                  doColumnar(lhsBool, f)
+              }
+            }
+            GpuColumnVector.from(finalRet, dataType)
+          }
+        }
+      }
+    }
+  }
 }
 
-case class GpuOr(left: Expression, right: Expression) extends CudfBinaryOperator with Predicate {
-  override def inputType: AbstractDataType = BooleanType
+case class GpuOr(left: Expression, right: Expression) extends CudfBinaryPredicateWithSideEffect {
 
   override def symbol: String = "||"
 
   override def sqlOperator: String = "OR"
 
   override def binaryOp: BinaryOp = BinaryOp.NULL_LOGICAL_OR
   override def astOperator: Option[BinaryOperator] = Some(ast.BinaryOperator.NULL_LOGICAL_OR)
+
+  override def columnarEvalWithSideEffects(batch: ColumnarBatch): Any = {
+    val leftExpr = left.asInstanceOf[GpuExpression]
+    val rightExpr = right.asInstanceOf[GpuExpression]
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuColumnVector.from(batch)) { tbl =>
+      withResource(GpuExpressionsUtils.columnarEvalToColumn(leftExpr, batch)) { lhsBool =>
+        if (isAllTrue(lhsBool)) {
+          withResource(GpuScalar.from(true, dataType)) { trueScalar =>
+            GpuColumnVector.from(trueScalar, lhsBool.getRowCount.toInt, BooleanType)
+          }
+        } else {
+          // Invert the lhs.
+          withResource(lhsBool.getBase.unaryOp(UnaryOp.NOT)) { lhsInverted =>
+            // Replace nulll values
+            val lhsNoNulls = withResource(Scalar.fromBool(true)) { trueScalar =>
+              lhsInverted.replaceNulls(trueScalar)
+            }
+            // filter batch based on the inverse of the LHS
+            withResource(lhsNoNulls) { _ =>
+              val rEval = withResource(filterBatch(tbl, lhsNoNulls, colTypes)) { leftInverseBatch =>
+                rightExpr.columnarEval(leftInverseBatch)
+              }
+              val finalRet = withResourceIfAllowed(rEval) { _ =>
+                (rEval) match {
+                  case (f: GpuColumnVector) =>
+                    withResource(gather(lhsNoNulls, f)) { combinedVector =>
+                      doColumnar(lhsBool, GpuColumnVector.from(combinedVector, dataType))
+                    }
+                  case (t: GpuColumnVector, f: GpuScalar) =>
```

Comment:
This seems incorrect and I think this was intended to be `case (f: GpuScalar)` ?

Reply:
This path is not covered by the current tests. Would we ever hit this path? I don't think that scalar expressions can have side-effects and we would only call this method if the RHS has side-effects.